### PR TITLE
Select parser

### PR DESCRIPTION
**Pull request by novln**
Not working at the moment, highly unstable.

### CODE REVIEW

File: parser/analyzer_test.go [hunk 1]
```diff
@@ -0,0 +1,218 @@
...
+				Table:     true,
+			},
+			Expected: parser.AnalyzerResult{
+				Operation: "SELECT",
+				Table:     "foobar",
+			},
+		},
+		{
+			// Scenario #2
+			Query: "SELECT a, b, c FROM foobar AS example",
```
Comment (thoas) [inline parser/analyzer_test.go]: missing scenario: `SELECT f.a, f.b, f.c FROM foobar f`

File: parser/analyzer.go [hunk 2]
```diff
@@ -0,0 +1,217 @@
...
+		case token.Select:
+			if mode == token.Illegal {
+				mode = e.Type
+			}
+			if !option.Operation {
+				continue
+			}
+
+			result.Operation = e.Type.String()
+			if !option.Continue(result) {
```
Comment (thoas) [inline parser/analyzer.go]: maybe reverse this thing using `result.Filled()` to keep an option as an option without operation

File: parser/parser_test.go [hunk 3]
```diff
@@ -0,0 +1,1109 @@
...
+		is.True(query.Offset.IsEmpty())
+		is.True(query.Suffix.IsEmpty())
+
+		ctx := &types.RawContext{}
+		result.Write(ctx)
+		is.Equal(input, ctx.Query())
+	}
+	{
+		input := fmt.Sprint(
+			"SELECT a, b, c FROM test2 INNER JOIN test4 ON (test4.gid = test2.id AND test4.d = test2.d) ",
```
Comment (thoas) [inline parser/parser_test.go]: it's the best case scenario, you don't retrieve parenthesis around the `join` operator everywhere

File: parser/parser_test.go [hunk 4]
```diff
@@ -0,0 +1,1109 @@
...
+		is.True(query.Limit.IsEmpty())
+		is.True(query.Offset.IsEmpty())
+		is.True(query.Suffix.IsEmpty())
+
+		ctx := &types.RawContext{}
+		result.Write(ctx)
+		is.Equal(input, ctx.Query())
+	}
+	{
+		input := "SELECT id FROM table WHERE ((id = 1) AND (slug = 'foo'))"
```
Comment (thoas) [inline parser/parser_test.go]: same for this

File: parser/analyzer.go (hunk 2 above)
Comment (novln) [reply parser/analyzer.go]: The option is mandatory, but I could change that to `result.Finished(option)` if you prefer.

Comment (novln) [reply parser/analyzer.go]: Maybe option is a bad wording and we should change that to `query` or anything else.

File: lexer/lexer.go [hunk 5]
```diff
@@ -157,13 +158,42 @@ func (l *Lexer) getDelimiterToken() (token.Token, bool) { // nolint: gocyclo
...
 		return token.Token{}, false
 	}
 }
 
+func (l *Lexer) unwrapDelimiter(delimiter rune, handler func() token.Token) token.Token {
+	if l.current() != delimiter {
+		l.error(errors.Errorf("identifier not started by %c", delimiter))
+		return token.Token{
+			Type:  token.Illegal,
+			Value: fmt.Sprintf("%c", l.current()),
```
Comment (novln) [inline lexer/lexer.go]: `Value: string(l.current())`

**Comment by yansal** [comment]
It would be cool if this parser were able to parse postgresql `SET` queries, see (https://www.postgresql.org/docs/current/static/sql-set.html)

**Comment by thoas** [comment]
closing this issue, since we are not using it.

**Pull request closed without merging**